### PR TITLE
Enable automatic @slowTest detection based on previous nightly's S3 stats

### PR DESCRIPTION
**Pull request by janeyx99**
This is a draft that can start discussion on how we want to automate @slowTest detection. Currently, this code just pulls nightly stats from S3 (for every test file) and uses them to detect whether a test case should be considered slow. The current requirement is that it must be a passed test case greater than 60 seconds. Pretty intuitive so far.

* The relative importing isn't working because tools is outside of torch, but the common_utils.py file getting used is in torch. I am not sure how to fix this.

There are several concerns:
1. There is duplicated code from test/run_test.py.
2. The data we use is per CI job, whereas we probably want a global bank of slow tests ALL jobs should avoid unless the job is marked as slow.
3. There is no real persistence. Consider the following situation: Run #1 reads the stats and marks test A as slow. test A is skipped for some runs. A day later, because the test was skipped, Run #100 reads the stats and does NOT mark test A as slow since it did not find a passing test A > 60s in the stats. Test A is now enabled again. And this goes on and off forever. 

My proposal/next steps: 
1. We host a list of slow tests on S3. We have a periodic master CI job (daily, let's say) that updates this list based on the test times of every CI job run on master. This would solve #2 and #3 from above. 
2. We have code to pull from this list during CI build jobs so that they can be used for spawned test jobs.
3. Oh hey wait a minute, #2 sounds like what we're doing for sharding (we generate a test-file-times file during a build job). So we can move the code for that in run_test and the code for #2 into a script in tools. The script in tools/ would export files containing S3 stats we care about. For sharding, it would be test file times. For slowTests, it would be test-case-times. This would solve problem #1. 

Any thoughts and suggestions on my proposal?

### CODE REVIEW

**Comment by walterddr** [comment]
for the first step I think we can maintain the import of tools in one location instead of in both run_test.py & common_utils.py
I think only import tools/* in run_test.py is better idea since run_test.py is executable python script. and in order for common_utils.py (or other internal test tooling) to access s3 data or other external data. we can have run_test.py to write those into files like how we did it for the shard?

for the persistency issue i think it is a good idea to persist it somewhere (S3, etc) and have a master branch chronos job to update the list. (so thumbs up for the option 3 in follow up steps)

File: torch/testing/_internal/common_utils.py [hunk 1]
```diff
@@ -55,6 +57,13 @@
 import torch.backends.mkl
 from enum import Enum
 
+try:
+    sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), "../../.."))
+    from tools.stats_utils.s3_stat_parser import (get_S3_bucket_readonly, HAVE_BOTO3, Report, Version2Report)
+except ImportError:
+    print("Unable to import s3_stat_parser from tools. Running without S3 stats...")
+    HAVE_BOTO3 = False
```
Comment (walterddr) [inline torch/testing/_internal/common_utils.py:65]: I am not sure this would work well here, I only tried this when the script is `__main__`

Comment (janeyx99) [reply torch/testing/_internal/common_utils.py:65]: Yea, this doesn't work because the __file__ is the one in site_packages.

**Comment by janeyx99** [comment]
> for the first step I think we can maintain the import of tools in one location instead of in both run_test.py & common_utils.py
> I think only import tools/* in run_test.py is better idea since run_test.py is executable python script. and in order for common_utils.py (or other internal test tooling) to access s3 data or other external data. we can have run_test.py to write those into files like how we did it for the shard?

I agree.

> for the persistency issue i think it is a good idea to persist it somewhere (S3, etc) and have a master branch chronos job to update the list. (so thumbs up for the option 3 in follow up steps)

Oh, the next steps are all one proposal (not three separate options). They all tie with each other.

**Comment by walterddr** [comment]
yes. sorry. i meant the option 3 that's actually the key to solve the problem

**Comment by janeyx99** [comment]
> yes. sorry. i meant the bulletin 3 that's explaining the key of the solution (s/option/bulletin/g because using "#" will accidentally mention other objects in github)

oh yea ooops

**Comment by samestep** [comment]
Perhaps a silly question: why does concern 3 occur? When we mark a test as slow, doesn't it still get run in some configurations? so why would the 100th test (the next day) not find any runs at all containing that slow test?

**Comment by janeyx99** [comment]
> Perhaps a silly question: why does concern 3 occur? When we mark a test as slow, doesn't it still get run in some configurations? so why would the 100th test (the next day) not find any runs at all containing that slow test?

Because the current configuration has it pulling per job, so even if the test gets run in another job, it doesn't see that.

**Comment by janeyx99** [comment]
This feature is now in PyTorch without issues (I think) so will close this PR now

**Pull request closed without merging**